File: java/java/src/main/java/org/signal/libsignal/crypto/Aes256GcmDecryption.java
//
// Copyright 2021 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.crypto;

import org.signal.client.internal.Native;
import org.whispersystems.libsignal.InvalidKeyException;

public class Aes256GcmDecryption {
  public static final int TAG_SIZE_IN_BYTES = 16;

  private long handle;

  public Aes256GcmDecryption(byte[] key, byte[] nonce, byte[] associatedData) throws InvalidKeyException {
    this.handle = Native.Aes256GcmDecryption_New(key, nonce, associatedData);
  }

  @Override
  protected void finalize() {
    Native.Aes256GcmDecryption_Destroy(this.handle);
  }

  public void decrypt(byte[] plaintext) {
    Native.Aes256GcmDecryption_Update(this.handle, plaintext, 0, plaintext.length);
  }

  public void decrypt(byte[] plaintext, int offset, int length) {
    Native.Aes256GcmDecryption_Update(this.handle, plaintext, offset, length);
  }

  public boolean verifyTag(byte[] tag) {
    boolean tagOk = Native.Aes256GcmDecryption_VerifyTag(this.handle, tag);
    Native.Aes256GcmDecryption_Destroy(this.handle);
    this.handle = 0;
    return tagOk;
  }

}

File: swift/Sources/SignalClient/Aes256GcmDecryption.swift
//
// Copyright 2020 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import SignalFfi
import Foundation

public class Aes256GcmDecryption: ClonableHandleOwner {
    public init<KeyBytes, NonceBytes, AssociatedDataBytes>(
        _ key: KeyBytes,
        _ nonce: NonceBytes,
        _ associated_data: AssociatedDataBytes) throws
        where KeyBytes: ContiguousBytes,
              NonceBytes: ContiguousBytes,
              AssociatedDataBytes: ContiguousBytes {

        let handle: OpaquePointer? = try key.withUnsafeBytes { keyBytes in
            try nonce.withUnsafeBytes { nonceBytes in
                try associated_data.withUnsafeBytes { adBytes in
                    var result: OpaquePointer?
                    let error = signal_aes256_gcm_decryption_new(&result,
                                                                 keyBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                                 keyBytes.count,
                                                                 nonceBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                                 nonceBytes.count,
                                                                 adBytes.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                                 adBytes.count)
                    try checkError(error)
                    return result
                }
            }
        }
        super.init(owned: handle!)
    }

    internal override class func destroyNativeHandle(_ handle: OpaquePointer) -> SignalFfiErrorRef? {
        return signal_aes256_gcm_decryption_destroy(handle)
    }

    public func decrypt<MutableBytes: ContiguousMutableBytes>(_ bytes: inout MutableBytes) throws {
        try bytes.withUnsafeMutableBytes {
            let error = signal_aes256_gcm_decryption_update(nativeHandle,
                                                            $0.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                            $0.count,
                                                            0,
                                                            UInt32($0.count))
            try checkError(error)
        }
    }

    public func decrypt<MutableBytes: ContiguousMutableBytes>(
        _ bytes: inout MutableBytes, _ offset: UInt32,
        _ length: UInt32) throws {
        try bytes.withUnsafeMutableBytes {
            let error = signal_aes256_gcm_decryption_update(nativeHandle,
                                                            $0.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                            $0.count,
                                                            offset,
                                                            length)
            try checkError(error)
        }
    }

    public func verifyTag<Bytes: ContiguousBytes>(_ tag: Bytes) throws -> Bool{
        try tag.withUnsafeBytes {
            var result = false
            let error = signal_aes256_gcm_decryption_verify_tag(&result,
                                                                nativeHandle,
                                                                $0.baseAddress?.assumingMemoryBound(to: UInt8.self),
                                                                $0.count)
            try checkError(error)
            return result
        }
    }
}
